fix(l1): ignore transactions from peers before sync

avilagaston9 opened the pull request:
**Motivation**

We are currently disconnecting from peers for trying to process their transactions while not yet synced.

**Description**

This PR provides a hotfix that discards any incoming transaction from P2P until we are synced.
- Adds a `synced` flag to our storage.
- Sets the flag to `false` when starting the node and every time the sync is triggered.
- Sets the flag to `true` on every successful forkchoice update.
- Checks the flag before handling `Transactions`, `NewPooledTransactionHashes` and `PooledTransactions` messages.
- Adds `eth_syncing` hive test to the CI.
- Logs `MempoolErrors` instead of cutting the connection.

Closes #1686 
Closes #179
Closes #337

mpaulucci commented:
Can you update https://github.com/lambdaclass/ethrex/blob/main/crates/networking/rpc/eth/client.rs#L30 please?

fmoletta commented on crates/networking/p2p/sync.rs:
```diff
@@ -73,12 +73,20 @@ impl SyncManager {
                     "Sync finished, time elapsed: {} secs",
                     start_time.elapsed().as_secs()
                 );
+                if let Err(e) = store.update_sync_status(true) {
```
The info above is misleading, it doesn't refer to the node being synced but that a sync cycle was completed. A snap sync normally spawns across multiple cycles as the pivot moves

avilagaston9 replied:
Thanks 🙌🏻. Moved to `apply_fork_choice` in [7709bc6](https://github.com/lambdaclass/ethrex/pull/1754/commits/7709bc6ec45ff8b16774e22190081989891d2adf)

avilagaston9 commented:
> Can you update https://github.com/lambdaclass/ethrex/blob/main/crates/networking/rpc/eth/client.rs#L30 please?
> @mpaulucci 

Updated and added `eth_syncing` Hive test to CI.

mpaulucci commented on crates/blockchain/fork_choice.rs:
```diff
@@ -134,6 +134,9 @@ pub fn apply_fork_choice(
         store.update_safe_block_number(safe.header.number)?;
     }
     store.update_latest_block_number(head.number)?;
+    if let Err(e) = store.update_sync_status(true) {
```
Does it make sense to deal with an error here different than the other cases? Why not `store.update_sync_status(true)?`?

avilagaston9 replied:
You are right! Addressed in [7b9eff5](https://github.com/lambdaclass/ethrex/pull/1754/commits/7b9eff521a9aae278b76ace46bccbb4893dbc674).

mpaulucci commented on crates/networking/p2p/rlpx/connection.rs:
```diff
@@ -352,6 +352,7 @@ impl<S: AsyncWrite + AsyncRead + std::marker::Unpin> RLPxConnection<S> {
         sender: mpsc::Sender<Message>,
     ) -> Result<(), RLPxError> {
         let peer_supports_eth = self.capabilities.contains(&CAP_ETH);
+        let synced = self.storage.synced()?;
```
nit: `is_synced` makes it clearer that it's a boolean response

avilagaston9 replied:
Updated in [c0564e8](https://github.com/lambdaclass/ethrex/pull/1754/commits/c0564e879b3d361fb3fd9d002913bef3904cd2da)!

mpaulucci commented on crates/storage/store/engines/in_memory.rs:
```diff
@@ -50,6 +50,7 @@ struct ChainData {
     // TODO (#307): Remove TotalDifficulty.
     latest_total_difficulty: Option<U256>,
     pending_block_number: Option<BlockNumber>,
+    synced: Option<bool>,
```
can't we just have a bool that starts on false? Don't see the advantage of an Option

avilagaston9 replied:
Addressed in [35a072b](https://github.com/lambdaclass/ethrex/pull/1754/commits/35a072b4585d11abf380c6ad873831d966fea3ea)!

mpaulucci reviewed (approved):
some nits but looking great!

Arkenan commented on crates/blockchain/fork_choice.rs:
```diff
@@ -134,6 +134,7 @@ pub fn apply_fork_choice(
         store.update_safe_block_number(safe.header.number)?;
     }
     store.update_latest_block_number(head.number)?;
+    store.update_sync_status(true)?;
```
Are we sure this is the way to determine that the sync is finished? We can receive a fork choice updated engine message pretty fast down the line even if we didn't sync yet.

fmoletta replied:
That's the point, if we receive a fork choice and we are not synced yet then we would return a `Syncing` response and continue to the sync, we would not apply the fork choice so this is the correct place to check wether we are currently synced or not

Arkenan commented on crates/storage/store/storage.rs:
```diff
@@ -999,6 +1000,13 @@ impl Store {
             .is_some())
     }
+
+    pub fn is_synced(&self) -> Result<bool, StoreError> {
+        self.engine.is_synced()
+    }
```
Maybe we should match against the option and in the case where the syncing variable is not set, we assume it's false? In this way we also make sure that when we just start the node we are not in a synced state until we are sure we are.

It's either this or setting it to false on node startup.

avilagaston9 replied:
If we omit setting it to false on node startup, could it remain set to true from a previous run?

fmoletta commented on crates/storage/store/engines/libmdbx.rs:
```diff
@@ -514,6 +514,15 @@ impl StoreEngine for Store {
     }
+    fn is_synced(&self) -> Result<bool, StoreError> {
+        match self.read::<ChainData>(ChainDataIndex::IsSynced)? {
+            None => Err(StoreError::Custom("Sync status not found".to_string())),
```
Couldn't we just asume None = false here? This would mean we can also skip setting it to false upon startup

avilagaston9 replied:
IMO, we shouldn't omit setting it to false upon startup as it may remain set to true from a previous run.

fmoletta replied:
Makes sense

fmoletta reviewed (approved):
🚀
